File: res/base/scripts/screens/options_screen.lua
class "OptionsScreen" (Screen)

local kVSyncOn = "VSync On"
local kVSyncOff = "VSync Off"
local kWindowBorderOn = "Window Border On"
local kWindowBorderOff = "Window Border Off"
local kFullscreenOn = "Fullscreen On"
local kFullscreenOff = "Fullscreen Off"

local kResolutions = {
    { 800, 600 },
    { 1024, 768 },
    { 1280, 720 },
    { 1360, 768 },
    { 1600, 1900 },
    { 1920, 1080 }
}

function OptionsScreen:init()
    local saved_config = UserSave:getConfig()

    self.is_vsync = UserSave:isVSync()
    self.is_borderless = UserSave:isBorderless()
    self.is_fullscreen = UserSave:isFullscreen()
    self.res_index = 1
    self.res_buttons = {}

    local btn = Button("backBtn")
    btn:setText("Main Screen")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setColour("red")
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setAlignment("RIGHT|TOP", -64, 64)
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:addCallback("MouseUp_Left", self.onBackBtnClick, self)
    self:attach(btn)

    btn = Button("applyBtn")
    btn:setText("Apply")
    btn:setFont("system_15_fnt")
    btn:setRect(0, 0, 256, 64)
    btn:setAlignment("RIGHT|BOTTOM", -64, -64)
    btn:setTextAlignment("CENTER|MIDDLE")
    btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    btn:setColour("white")
    btn:addCallback("MouseUp_Left", self.onApplyBtnClick, self)
    self:attach(btn)

    self.vsync_btn = Button("vsyncBtn")
    self.vsync_btn:setText(self.is_vsync and kVSyncOn or kVSyncOff)
    self.vsync_btn:setFont("system_15_fnt")
    self.vsync_btn:setRect(0, 0, 256, 64)
    self.vsync_btn:setAlignment("LEFT|TOP", 64, 64)
    self.vsync_btn:setTextAlignment("CENTER|MIDDLE")
    self.vsync_btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    self.vsync_btn:setColour(self.is_vsync and "green" or "red")
    self.vsync_btn:addCallback("MouseUp_Left", self.onVsyncBtnClick, self)
    self:attach(self.vsync_btn)

    self.borderless_btn = Button("borderlessBtn")
    self.borderless_btn:setText(self.is_borderless and kWindowBorderOn or kWindowBorderOff)
    self.borderless_btn:setFont("system_15_fnt")
    self.borderless_btn:setRect(0, 0, 256, 64)
    self.borderless_btn:setAlignment("LEFT|TOP", 64, 128)
    self.borderless_btn:setTextAlignment("CENTER|MIDDLE")
    self.borderless_btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    self.borderless_btn:setColour(self.is_borderless and "green" or "red")
    self.borderless_btn:addCallback("MouseUp_Left", self.onBorderlessBtnClick, self)
    self:attach(self.borderless_btn)

    self.fullscreen_btn = Button("fullscreenBtn")
    self.fullscreen_btn:setText(self.is_fullscreen and kFullscreenOn or kFullscreenOff)
    self.fullscreen_btn:setFont("system_15_fnt")
    self.fullscreen_btn:setRect(0, 0, 256, 64)
    self.fullscreen_btn:setAlignment("LEFT|TOP", 64, 196)
    self.fullscreen_btn:setTextAlignment("CENTER|MIDDLE")
    self.fullscreen_btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
    self.fullscreen_btn:setColour(self.is_fullscreen and "green" or "red")
    self.fullscreen_btn:addCallback("MouseUp_Left", self.onFullscreenBtnClick, self)
    self:attach(self.fullscreen_btn)

    local screen_width = Engine.getScreenWidth()
    local screen_height = Engine.getScreenHeight()

    for i, res in ipairs(kResolutions) do
        local is_selected = screen_width == res[1] and screen_height == res[2]

        btn = Button()
        btn:setText(string.format("%ix%i", res[1], res[2]))
        btn:setFont("system_15_fnt")
        btn:setRect(0, 0, 256, 64)
        btn:addCallback("MouseUp_Left", self.onResolutionBtnClick, {self, btn})
        btn:setTextAlignment("CENTER|MIDDLE")
        btn:setAlignment("LEFT|TOP", 320, (i + 1) * 64)
        btn:setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr")
        btn:setColour(is_selected and "green" or "red")
        self:attach(btn)

        table.insert(self.res_buttons, btn)

        if (is_selected) then
            self.res_index = i
        end
    end
end

function OptionsScreen:onBackBtnClick()
    Screens.load("LoadingScreen", "MainScreen")
end

function OptionsScreen:onApplyBtnClick()
    UserSave:setVSync(self.is_vsync)
    UserSave:setBorderless(self.is_borderless)
    UserSave:setFullscreen(self.is_fullscreen)
    local res = kResolutions[self.res_index]
    UserSave:setScreenWidth(res[1])
    UserSave:setScreenHeight(res[2])
    UserSave:save()

    Engine.addCommand({id = "restart"})
end

function OptionsScreen:onVsyncBtnClick()
    self.is_vsync = not self.is_vsync
    if (self.vsync_btn) then
        self.vsync_btn:setText(self.is_vsync and kVSyncOn or kVSyncOff)
        self.vsync_btn:setColour(self.is_vsync and "green" or "red")
    end
end

function OptionsScreen:onBorderlessBtnClick()
    self.is_borderless = not self.is_borderless

    if (self.borderless_btn) then
        self.borderless_btn:setText(self.is_borderless and kWindowBorderOn or kWindowBorderOff)
        self.borderless_btn:setColour(self.is_borderless and "green" or "red")
    end
end

function OptionsScreen:onFullscreenBtnClick()
    self.is_fullscreen = not self.is_fullscreen
    if (self.fullscreen_btn) then
        self.fullscreen_btn:setText(self.is_fullscreen and kFullscreenOn or kFullscreenOff)
        self.fullscreen_btn:setColour(self.is_fullscreen and "green" or "red")
    end
end

function OptionsScreen.onResolutionBtnClick(params)
    local self = params[1]
    local sender = params[2]

    for i, btn in ipairs(self.res_buttons) do
        if (btn == sender) then
            btn:setColour("green")
            self.res_index = i
        else
            btn:setColour("red")
        end
    end
end

File: src/options_screen.cpp
#include "options_screen.h"

#include "system_tools.h"

namespace redrevolt
{
struct Resolution
{
    int width;
    int height;
};

const std::vector<Resolution> KResolutions = {
    {800, 600},
    {1024, 768},
    {1280, 720},
    {1360, 768},
    {1600, 1900},
    {1920, 1080}
};

OptionsScreen::OptionsScreen(const std::string & id)
    : stren::Screen(id)
    , m_vsyncBtn(nullptr)
    , m_borderlessBtn(nullptr)
    , m_fullscreenBtn(nullptr)
    , m_isVSync(false)
    , m_isBorderless(false)
    , m_isFullscreen(false)
    , m_resIndex(0)
{
    SystemTools * systemTools = new SystemTools("systemTools");
    attach(systemTools);

    stren::GetConfigCommand command;
    command.execute();

    m_isVSync = command.isVSync();
    m_isFullscreen = command.isFullscreen();
    m_isBorderless = command.isBorderless();

    stren::Button * btn = new stren::Button("backBtn");
    btn->setText("Main Screen");
    btn->setFont("system_15_fnt");
    btn->setColour("red");
    btn->setRect(0, 0, 256, 64);
    btn->setAlignment("RIGHT|TOP", -64, 64);
    btn->setTextAlignment("CENTER|MIDDLE");
    btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    btn->addCallback("MouseUp_Left", this, &OptionsScreen::onBackBtnClick);
    attach(btn);

    btn = new stren::Button("applyBtn");
    btn->setText("Apply");
    btn->setFont("system_15_fnt");
    btn->setColour("white");
    btn->setRect(0, 0, 256, 64);
    btn->setAlignment("RIGHT|BOTTOM", -64, -64);
    btn->setTextAlignment("CENTER|MIDDLE");
    btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    btn->addCallback("MouseUp_Left", this, &OptionsScreen::onApplyBtnClick);
    attach(btn);

    m_vsyncBtn = new stren::Button("vsyncBtn");
    m_vsyncBtn->setText(m_isVSync ? "VSync On" : "VSync Off");
    m_vsyncBtn->setColour(m_isVSync ? "green" : "red");
    m_vsyncBtn->setFont("system_15_fnt");
    m_vsyncBtn->setRect(0, 0, 256, 64);
    m_vsyncBtn->setAlignment("LEFT|TOP", 64, 64);
    m_vsyncBtn->setTextAlignment("CENTER|MIDDLE");
    m_vsyncBtn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    m_vsyncBtn->addCallback("MouseUp_Left", this, &OptionsScreen::onVSyncBtnClick);
    attach(m_vsyncBtn);

    m_borderlessBtn = new stren::Button("borderlessBtn");
    m_borderlessBtn->setText(m_isBorderless ? "Window Border On" : "Window Border Off");
    m_borderlessBtn->setColour(m_isBorderless ? "green" : "red");
    m_borderlessBtn->setFont("system_15_fnt");
    m_borderlessBtn->setRect(0, 0, 256, 64);
    m_borderlessBtn->setAlignment("LEFT|TOP", 64, 128);
    m_borderlessBtn->setTextAlignment("CENTER|MIDDLE");
    m_borderlessBtn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    m_borderlessBtn->addCallback("MouseUp_Left", this, &OptionsScreen::onBorderlessBtnClick);
    attach(m_borderlessBtn);

    m_fullscreenBtn = new stren::Button("m_fullscreenBtn");
    m_fullscreenBtn->setText(m_isFullscreen ? "Fullscreen On" : "Fullscreen Off");
    m_fullscreenBtn->setColour(m_isFullscreen ? "green" : "red");
    m_fullscreenBtn->setFont("system_15_fnt");
    m_fullscreenBtn->setRect(0, 0, 256, 64);
    m_fullscreenBtn->setAlignment("LEFT|TOP", 64, 196);
    m_fullscreenBtn->setTextAlignment("CENTER|MIDDLE");
    m_fullscreenBtn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
    m_fullscreenBtn->addCallback("MouseUp_Left", this, &OptionsScreen::onFullscreenBtnClick);
    attach(m_fullscreenBtn);

    const int screenWidth = stren::Engine::getScreenWidth();
    const int screenHeight = stren::Engine::getScreenHeight();

    size_t i(0);
    for (const Resolution & res : KResolutions)
    {
        const std::string resStr = std::to_string(res.width) + "x" + std::to_string(res.height);
        const bool isSelected = screenWidth == res.width && screenHeight == res.height;
        stren::Button * btn = new stren::Button();
        btn->setText(resStr);
        btn->setFont("system_15_fnt");
        btn->setTextAlignment("CENTER|MIDDLE");
        btn->setRect(0, 0, 256, 64);
        btn->setAlignment("LEFT|TOP", 320, (i + 1) * 64);
        btn->setColour(isSelected ? "green" : "red");
        btn->setSprites("up_btn_spr", "down_btn_spr", "over_btn_spr");
        btn->addCallback("MouseUp_Left", this, &OptionsScreen::onResolutionBtnClick);
        m_resButtons.push_back(btn);
        attach(btn);

        if (isSelected)
        {
            m_resIndex = i;
        }
        ++i;
    }
}

OptionsScreen::~OptionsScreen()
{
}

void OptionsScreen::onBackBtnClick(Widget * sender)
{
    stren::SwitchScreenCommand command;
    command.setScreen("LoadingScreen");
    command.setNextScreen("MainScreen");
    command.execute();
}

void OptionsScreen::onApplyBtnClick(Widget * sender)
{
    stren::UpdateConfigCommand configCommand;
    if (m_resIndex < KResolutions.size())
    {
        const Resolution & res = KResolutions[m_resIndex];
        configCommand.setScreenWidth(res.width);
        configCommand.setScreenHeight(res.height);
    }
    configCommand.setBorderless(m_isBorderless);
    configCommand.setFullscreen(m_isFullscreen);
    configCommand.setVSync(m_isVSync);
    configCommand.execute();

    stren::SerializeCommand serializeCommand;
    serializeCommand.execute();

    stren::RestartCommand restartCommand;
    restartCommand.execute();
}

void OptionsScreen::onVSyncBtnClick(Widget * sender)
{
    m_isVSync = !m_isVSync;
    if (m_vsyncBtn)
    {
        m_vsyncBtn->setText(m_isVSync ? "VSync On" : "VSync Off");
        m_vsyncBtn->setColour(m_isVSync ? "green" : "red");
    }
}

void OptionsScreen::onBorderlessBtnClick(Widget * sender)
{
    m_isBorderless = !m_isBorderless;
    if (m_borderlessBtn)
    {
        m_borderlessBtn->setText(m_isBorderless ? "Window Border On" : "Window Border Off");
        m_borderlessBtn->setColour(m_isBorderless ? "green" : "red");
    }
}

void OptionsScreen::onFullscreenBtnClick(Widget * sender)
{
    m_isFullscreen = !m_isFullscreen;
    if (m_fullscreenBtn)
    {
        m_fullscreenBtn->setText(m_isFullscreen ? "Fullscreen On" : "Fullscreen Off");
        m_fullscreenBtn->setColour(m_isFullscreen ? "green" : "red");
    }
}

void OptionsScreen::onResolutionBtnClick(Widget * sender)
{
    size_t i(0);
    for (stren::Button * btn : m_resButtons)
    {
        if (btn == sender)
        {
            btn->setColour("green");
            m_resIndex = i;
        }
        else
        {
            btn->setColour("red");
        }
        ++i;
    }
}
} // redrevolt
